File: srclib/v3/spss/srclib/combine.py
# combine function

import srclib
import re, spss, sys

def combine(name=None, dir=None, mode=None):

  proc = "combine"

  args = srclib.arguments(name, dir, mode, proc)  # get the arguments
  if not args:
    return 1
  path = args[0]
  datain = args[1]
  dataout = args[2]
  setup = args[3]

  var = None  # variable list
  m = re.search(r"(^|;)\s*var[\w]*\s+([^;]+);", setup, flags=re.IGNORECASE)  # supplied?
  if m:
    var = m.group(2).strip()
    var = var.strip()
    if not re.match(r"^[\w -]+$", var):  # check it
      sys.Exit("Invalid variable list.")  # invalid

  filenames = datain.strip()  # build the file names list
  filelist = []
  while filenames != "":  # loop through file names
    m = re.match(r'^"([^"]+)"(\s+.*)?$', filenames)  # double quotes
    if not m:
      m = re.match(r"^'([^']+)'(\s+.*)?$", filenames)  # single quotes
      if not m:
        m = re.match(r"^([^\s]+)(\s+.*)?$", filenames)  # no quotes
    if m.group(2):
      filenames = m.group(2).strip()  # trim the remaining names
    else:
      filenames = ""
    file = m.group(1).strip()  # extract the file name
    m = re.match(r"^[~\.\\/:\w -]+$", file)  # check it
    if not m:
      sys.Exit("Invalid datain.")  # invalid
    if not re.match(r"\.sav$", file, flags=re.IGNORECASE):  # add .sav
      file += ".sav"
    filelist.append(file)  # append it to the list

  #dataout
  if not re.match(r"\.sav$", dataout, flags=re.IGNORECASE):  # add .sav
    dataout += ".sav"

  beg = 0  # merge the files
  count = 0
  for file in filelist:  # loop through the input files
    if count > 0:  # not the first file
      cmd = "save outfile='{0}'.\n".format(dataout)  # save the intermediate file
      spss.Submit(cmd)
    cmd = "get file='{0}'.\n".format(file)  # get the input file
    spss.Submit(cmd)
    cmd = "compute OBS_ = $casenum.\n"  # compute the observation number
    cmd += "compute OBS_ = OBS_ + {0}.\n".format(beg)
    cmd += "execute.\n"
    spss.Submit(cmd)
    beg += spss.GetCaseCount()  # increment the case count
    if count > 0:  # not the first file
      cmd = "match files file=* /file='{0}' /by OBS_.\n".format(dataout)  # match the input and intermediate files
      spss.Submit(cmd)
    count += 1

  if var:  # variable list
    cmd = "save outfile='{0}' /keep={1}.\n".format(dataout, var)  # subset variables and save
  else:
    cmd = "save outfile='{0}'.\n".format(dataout)  # save the output file
  spss.Submit(cmd)

File: srclib/v3/R/combine.R

# function to combine R datasets

combine <- function(name="", dir="", mode="") {
  proc <<- "combine"
  msg <<- ""
  path <<- ""
  setup <<- ""
  test <<- ""
  datain <<- ""
  dataout <<- ""
  vars <<- as.null()
  arguments(name, dir, mode)  # get the arguments

  if (identical(msg, "")) {  # datain
    datain <<- gsub("(\\s+\"[^\"]+\"|\\s+'[^']+'|\\s+[^\\s]+)", ",\\1", datain)  # add commas
    datain <<- gsub("(\"|')", "", datain)  # remove quotes
    datain <<- gsub("(\\s*,\\s*)", ",", datain)  # remove spaces around commas
    datain <<- strsplit(datain, ",")  # convert to list
    datain <<- datain[[1]]
  }

  if (identical(msg, "")) {  # variable list
    if (grepl("(^|;)\\s*var\\s+[^;]+;", setup, ignore.case=TRUE)) {
      vars <<- sub("^(.*;)*\\s*var\\s+([^;]+);.*$", "\\2", setup, ignore.case=TRUE)  # variable names
      vars <<- sub("\\s\\s+", " ", sub("^\\s+", "", sub("\\s+$", "", vars)))  # trim
      if (!grepl("^[[:alnum:]_ -]+$", vars))  msg <<- "Invalid variable list."
      else {
        vars <<- unlist(strsplit(vars, " "))  # convert to vector
      }
    }
  }

  if (identical(msg, "")) {  # combine the files
    beg <<- 0
    dslist = lapply(datain, function(x) { dsread(x) })
    ds <<- Reduce(function(x, y) { merge(x, y, all=TRUE) }, dslist)

    if (!is.null(vars)) {  # variable list
      ds <<- ds[vars]  # subset
    }

    # save dataout
    if (!grepl("\\.rda$", dataout, ignore.case=TRUE)) dataout <- paste(dataout, "rda", sep=".")  # add the suffix
    save(ds, file=dataout)  # save
  }
}

dsread <- function(file) {  # read the data file
  type <- ""  # get the type
  if (grepl(".+\\.[[:alnum:]]+$", file)) type <- tolower(sub(".+\\.([[:alnum:]]+)$", "\\1", file))
  if (identical(type, "csv")) {  # csv
    ds <- read.csv(file)
  }
  else if (identical(type, "dta")) {  # dta
    library(foreign)
    ds <- read.dta(file)
  }
  else if (identical(type, "rda") || identical(type, "rdata")) {  # rda
    ds <- get(load(file))
  }
  else if (identical(type, "sav")) {  # sav
    library(foreign)
    ds <- read.spss(file, to.data.frame = TRUE)
  }
  else if (identical(type, "txt")) {  # txt
    ds <- read.delim(file)
  }
  else if (identical(type, "xpt")) {  # xpt
    library(foreign)
    ds <- read.xport(file)
  }
  else {  # default
    ds <- get(load(paste(file, "rda", sep=".")))
  }
  ds$OBS_ <- 1 : nrow(ds)  # add observation number in the combined dataset
  ds$OBS_ <- ds$OBS_ + beg
  beg <<- beg + nrow(ds)
  return (ds)
}

